### DISCUSSION TITLE
Using Hyper headers in Rocket 0.5

### DISCUSSION THREAD

**Discussion by barlindhaug** [rwf2/Rocket#1719] [Questions] [answered]
I've started migrating to rocket 0.5 and ran into an issue when using hyper headers.

```
.header(Header::new(
    hyper::header::CACHE_CONTROL.as_str(),
    "max-age=2678400",
))
```

Looking at [rocket::http::struct::Header::new](https://api.rocket.rs/v0.5-rc/rocket/http/struct.Header.html#method.new), I see 

> Constructs a new header. This method should be used rarely and only for non-standard headers. Instead, prefer to use the Into\<Header\> implementations of many types, including ContentType and all of the headers in http::hyper::header.

So my question is how I should use the `Into<Header>` implementation.

Previously on rocket 0.4 the code was
```
.header(CacheControl(vec![CacheDirective::MaxAge(2678400u32)]))
```

**Comment by jebrosen** [comment, marked as answer]
> So my question is how I should use the `Into<Header>` implementation.

Yes, a lot of these have been lost between 0.4 and 0.5-rc. That's now a mismatch in documentation, so this text will need to be updated or removed.

Rocket 0.4 was based on hyper 0.10, which included those typed headers, and rocket re-exported them. The typed headers were removed some time between then and hyper 0.13, and we haven't yet found a suitable replacement (see issue #1067). The most recent attempt was in #1535 using the `hyperx` crate, but the performance costs were deemed too high to go with that approach. Unfortunately, the next-best and more performant option, `headers`, has some usability issues; in particular several header types in the library are not actually constructible.
